File: Data Structures & Algorithms/valid-binary-search-tree/submission-4.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

 

class Solution {
    
    public boolean isValidBST(TreeNode root) {
        /**
        Given root of binary tree --> True if valid, otherwise false


        To be a valid BST
            Left subtree only contains nodes with keys less than node's key
            Right subtree only contains nodes with keys greater than node's key
            Both left and right subtrees are also BST
        */



        //First TODO: Null condition --> Null tree is a BST
        if(root == null) return true;
        //If left subtree less than root.val, and right val > root.val we would want to return true
    

        //if(root.left.val < root.val || root.right.val > root.val) return true;
        //This only checks the condition for the root, we need to iterate through the tree making sure
        //each left child of node is less than node, and each right child of node > node 

        return helper(root, Long.MIN_VALUE, Long.MAX_VALUE);


        
    }
    private boolean helper(TreeNode node, Long min, Long max){
        if(node == null) return true;
        if(node.val <= min) return false;
        if(node.val >= max) return false;
        return helper(node.left, min, (long) node.val) && helper(node.right, (long) node.val, max);
        


        
    }
}

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        '''
            given root of a binary search tree and an integer k, return the kth smallest value in the tree

            - left subtree of every node contains onlt nodes with kets less than the node's key
            - right subtree of every node contains omnly nodes greater than the node's key

            both the left and right subtrees are also binary search trees

                root = [2,1,3], k = 1

                return the smallest value in the tree, since its a bst go left, since its k = 1 it wouold be the left most value

                root = [4,3,5,2,null], k = 4
                would be the largest value since theres oly 4 values

                we need to traverse the tree comparing the values


                inorder traversal:
                - left, node, right
             
            inorder = 2, 3, 4, 5, 7, 8
            left node right



        '''
        count = 0
        ans = 0
        def dfs(node):
            nonlocal count
            nonlocal ans
            if not node:
                return
            dfs(node.left)
            #processing root
            count += 1
            if count == k:
                ans = node.val
                return
            return dfs(node.right)
    
        dfs(root)
        return ans

            

            

            